packages/hardhat-core/src/internal/core/vars/vars-manager-setup.ts

```diff
@@ -0,0 +1,74 @@
+import debug from "debug";
+import { VarsManager } from "./vars-manager";
+
+const log = debug("hardhat:core:vars:varsManagerSetup");
+
+// This class is ONLY used when collecting the required and optional vars that have to be filled by the user
+export class VarsManagerSetup extends VarsManager {
+  private readonly _requiredVars: Set<string>;
+  private readonly _optionalVars: Set<string>;
+
+  constructor(varsFilePath: string) {
+    log("Creating a new instance of VarsManagerSetup");
+
+    super(varsFilePath);
+    this._requiredVars = new Set();
+    this._optionalVars = new Set();
+  }
+
+  // During setup we ignore env variables, so this function is the same as 'has'
+  public hasWithEnvVars(key: string): boolean {
+    return this.has(key);
+  }
+
+  // During setup we ignore env variables, so this function is the same as 'get'
+  public getWithEnvVars(
+    key: string,
+    defaultValue?: string | undefined
+  ): string | undefined {
+    return this.get(key, defaultValue);
+  }
+
+  // Checks if the key exists, and updates optional sets accordingly.
+  // Required vars have priority over optional vars.
+  public has(key: string): boolean {
+    log(`function 'has' called with key '${key}'`);
+
+    if (super.has(key)) return true;
+
+    if (!this._requiredVars.has(key)) this._optionalVars.add(key);
```

Comment:
Personal preference: I'm not a fan of ignoring the curly braces `{ }` in code blocks.